component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java

```diff
@@ -0,0 +1,305 @@
+            log.error("Failed calling Moesif microservice. Organization ID cannot be empty");
+            return null;
+        }
+        final URL obj;
+        try {
+            obj = new URL(url);
+        } catch (MalformedURLException ex) {
+            log.error("Failed calling Moesif microservice. Attempted to call url: {}", url.replaceAll("[\r\n]", ""),
+                    ex.getMessage().replaceAll("[\r\n]", ""));
+            return null;
+        }
+        String authHeaderValue = getAuthHeader(gaAuthUsername, gaAuthPwd);
+        HttpsURLConnection con = null;
+        try {
+            con = (HttpsURLConnection) obj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("Content-Type", MoesifMicroserviceConstants.CONTENT_TYPE);
+            con.setRequestProperty("Authorization", authHeaderValue);
+            con.setReadTimeout(MoesifMicroserviceConstants.REQUEST_READ_TIMEOUT);
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpsURLConnection.HTTP_OK) {
+                try (BufferedReader in = new BufferedReader(
+                        new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8));) {
+
+                    String inputLine;
+
+                    while ((inputLine = in.readLine()) != null) {
+                        response.append(inputLine);
+                    }
+                    updateMoesifKey(response.toString());
+                    return response.toString();
+                }
+            } else if (responseCode >= 400 && responseCode < 500) {
+                log.error("Getting {} from the microservice.", responseCode);
+                return null;
+            } else {
+                throw new APICallException("Getting " + responseCode + " from the microservice and retrying.");
+            }
+        } finally {
+            if (con != null) {
+                con.disconnect();
+            }
+        }
+    }
+
+    private String getAuthHeader(String gaAuthUsername, char[] gaAuthPwd) {
+        String auth = gaAuthUsername + ":" + Arrays.toString(gaAuthPwd);
+        String encodedAuth = Base64.getEncoder().encodeToString(auth.getBytes(StandardCharsets.UTF_8));
+        String authHeaderValue = "Basic " + encodedAuth;
+        return authHeaderValue;
+    }
+
+    private void updateMoesifKey(String response) {
+        Gson gson = new Gson();
+        String json = response;
+        MoesifKeyEntry newKey = gson.fromJson(json, MoesifKeyEntry.class);
+        String orgID = newKey.getOrganization_id();
+        String moesifKey = newKey.getMoesif_key();
+        orgIDMoesifKeyMap.put(orgID, moesifKey);
+        moesifKeyClientMap.put(moesifKey, new MoesifAPIClient(moesifKey));
+    }
+
+    private void updateMap(String response) {
+        Gson gson = new Gson();
+        String json = response;
+
+        Type collectionType = new TypeToken<Collection<MoesifKeyEntry>>() {
+        }.getType();
```

Comment:
```
Type collectionType = new CustomType().getType();
```

and define new static inner class 
```
    static class CustomType extends TypeToken<Collection<MoesifKeyEntry>> {
        // nothing to add
    }
    ```

Reply:
Decided to resolve this bug by excluding associated findbug.